[DOC] Cruise Control documentation updates

laidan6000 opened the pull request:
### Type of change

- Documentation

### Description

RESOLVES ISSUES #3094 and #3090 

This pull request contains various changes to the Cruise Control chapter of the documentation.

* New title: Cruise Control for cluster rebalancing
* Added a new paragraph at the start explaining what Cruise Control is and a brief explanation of what is NOT supported
* Added an example: `Kafka` configuration for hard optimization goals
* Removed some redundant / ambiguous information
* Removed two references to the `CpuCapacityGoal` and the `CpuUsageDistributionGoal`, which are not yet supported
* Linked to the "Capacity configuration" section in a couple of places
* Updated the procedures as a result of testing
* Edited for consistency, e.g. "rebalance proposal" replaced with "optimization proposal" in a couple of places.
* Edited the "Cruise Control configuration" module:
** Make it consistent with the content on optimization goals and optimization proposals 
** Attempt to simplify the explanation of "Auto-created topics and log compaction"
* Changed links to cross-references when referring to a specific anchor

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [X] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [X] Reference relevant issue(s) and close them after merging - 
- [ ] Update CHANGELOG.md

ppatierno commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -30,66 +30,91 @@ For more information on each optimization goal, see link:https://github.com/link
 Cruise Control will ignore this goal if doing so enables all the configured hard goals to be met.
 
-In Cruise Control, six of the link:#master-goals[master optimization goals] are preset as hard goals:
+In Cruise Control, six of the xref:#master-goals[master optimization goals] are preset as hard goals:
```
do we need to use a specific number here (six)? If something changes in the future with less or more hard goals I think we will totally forget to update this number. I am pretty sure :-) Maybe just using "some" ... or your call ;-)

laidan6000 replied:
Will change to say "the following master optimization goals".

Also, there are five not six. D'oh!

ppatierno commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -30,66 +30,91 @@ For more information on each optimization goal, see link:https://github.com/link
+RackAwareGoal; ReplicaCapacityGoal; DiskCapacityGoal; NetworkInboundCapacityGoal; NetworkOutboundCapacityGoal; ReplicaDistributionGoal; PotentialNwOutGoal; DiskUsageDistributionGoal; NetworkInboundUsageDistributionGoal; NetworkOutboundUsageDistributionGoal; TopicReplicaDistributionGoal; LeaderReplicaDistributionGoal; LeaderBytesInDistributionGoal; PreferredLeaderElectionGoal
 
-Six of these goals are preset as link:#hard-soft-goals[hard goals].
+Six of these goals are preset as xref:#hard-soft-goals[hard goals].
```
Again a number here (Six) I would avoid that as I mentioned before

tomncooper commented on documentation/assemblies/cruise-control/assembly-cruise-control-concepts.adoc:
```diff
@@ -3,11 +3,16 @@
+Cruise Control is an open source system for automating Kafka operations, including monitoring cluster workload, rebalancing a cluster based on predefined constraints, and detecting and fixing anomalies, among others. 
+It consists of four main components--the Load Monitor, the Analyzer, the Anomaly Detector, and the Executor--and a REST API for all interactions.
+{ProductName} utilizes the REST API to support the cluster rebalancing features of Cruise Control.
+Other features, such as anomaly detection, Slack notifications, and changing the topic replication factor, are not supported.
```
Could we change this to "not currently supported"? We will most likely be add topic replication in the future, not sure about slack notifications though? Maybe change slack notifications to something ore general like "anomaly notifications"

tomncooper commented on documentation/modules/cruise-control/proc-generating-optimization-proposals.adoc:
```diff
@@ -49,6 +49,8 @@ spec:
     - ReplicaCapacityGoal
 ----
++
+In the above example, if `RackAwareGoal` and `ReplicaCapacityGoal` are configured as the only master optimization goals (in `Kafka.spec.cruiseControl.config.goals`), you must add `skip_hard_goal_check=true` to the Cruise Control deployment configuration. Otherwise an error will occur. 
```
So I don't really understand this point. The issue would come if the configured `hard.goals` had more than those two goals in? If master goals had these two in and so did `hard.goals` then you would be fine. 

This highlights another issue that we should highlight that if a user changes master goals they should also make sure that they make sure `hard.goals` is still a subset of the new master goals or strange errors will occur (in the CC release we are using, the latest release adds more checks that would fail to start if the user messed this up).

im-konge replied:
This was issue that I mentioned, 'cause I was curious about it -> if you set for example these two `hard.goals` and then apply the KafkaRebalance, again with these two `hard.goals` the error will occur -> it shows something like `tried to find [all hard goals] in [rackAwareGoal, replicaCapacityGoal]` or something like that.

im-konge replied:
And it's because of me that @laidan6000 add this point, because I found it when I was testing the documentation.

tomncooper replied:
The issue is that the default `hard.goals` list has 5 (I think) goals in it. So if you set master goals to a list that does not include all those `hard.goals` then you will fail every rebalance (unless you set skip hard goals) and every periodic cached proposal calculation will fail as well.

This will be fixed in the next strimzi release becasue I filed an issue for it and it has been fixed in the latest release and we can add some additional sanity checks as well if we need to. 

**For the current release** we need to document that the user **must** make sure that if they change the master goals that they also make sure that `hard.goals` is a subset of the master goals they set.

im-konge replied:
Okay, thanks for the explanation and the issue you created for CC team. I tested it like (because I was) a customer and it wasn't clear for me for the first look.

So from my POV it would be nice to have some note about what you mentioned.

laidan6000 replied:
I removed the sentence referred to above in line 53.

I added the following note to the master optimization goals section:

**NOTE:** If you change the inherited master optimization goals, you must ensure that the hard goals, if configured in the `hard.goals` property in `Kafka.spec.cruiseControl.config`, are a subset of the master optimization goals that you configured. Otherwise, errors will occur when generating optimization proposals.

tomncooper replied:
Good that you have added the note, but there is not enough context in this sentence for it to be meaningful. For the sentence to be true, a user would have had to break the rule we just outlined in the note in the master goals section? I think it would be better to remove this sentence all together.

laidan6000 replied:
OK. I have actually removed the sentence. Glad you're happy with the new NOTE.

tomncooper commented on documentation/modules/cruise-control/ref-cruise-control-configuration.adoc:
```diff
@@ -57,11 +57,13 @@ spec:
 Cruise Control uses _capacity limits_ to determine if optimization goals for broker resources are being broken.
-An optimization will fail if a goal is broken, preventing the optimization from being used to generate an optimization proposal for balanced partition assignments.
-For example, an optimization that would cause a broker to exceed its CPU capacity would not be used if the `CpuCapacityGoal` is set as a hard goal.
+An optimization will fail if a goal is broken, preventing the optimization from being used to generate an optimization proposal.
```
"If a **hard** goal is broken"

PaulRMellor commented on documentation/assemblies/cruise-control/assembly-cruise-control-concepts.adoc:
```diff
@@ -3,11 +3,16 @@
+You can deploy {CruiseControlProject} to your {ProductName} cluster and access its _cluster rebalancing_ feature through custom resources.
 
-Example YAML files for deploying Cruise Control and setting optimization goals are provided in `examples/cruise-control/`.
+Cruise Control is an open source system for automating Kafka operations, including monitoring cluster workload, rebalancing a cluster based on predefined constraints, and detecting and fixing anomalies, among others. 
```
Not sure we need _among others_ if we say _including_

laidan6000 replied:
Good point. Changed to "such as" and removed "among other".

PaulRMellor commented on documentation/assemblies/cruise-control/assembly-cruise-control-concepts.adoc:
```diff
@@ -3,11 +3,16 @@
+You can deploy {CruiseControlProject} to your {ProductName} cluster and access its _cluster rebalancing_ feature through custom resources.
 
-Example YAML files for deploying Cruise Control and setting optimization goals are provided in `examples/cruise-control/`.
+Cruise Control is an open source system for automating Kafka operations, including monitoring cluster workload, rebalancing a cluster based on predefined constraints, and detecting and fixing anomalies, among others. 
```
Do we want to list things we don't support here like anomaly detection? Maybe we could  be clear in this intro what is supported. _Strimzi currently supports the <> features of Cruise Control_

laidan6000 replied:
Right. It's not that helpful to start with an explanation what we don't support.

I'll add a list of supported features as you suggest, and relegate the not supported features to a note.

PaulRMellor commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -9,7 +9,7 @@ To rebalance a Kafka cluster, Cruise Control uses optimization goals to generate
 Optimization goals are constraints on workload redistribution and resource utilization across a Kafka cluster.
 With a few exceptions, {ProductName} supports all the optimization goals developed in the Cruise Control project.
-These are as follows, in descending priority order:
+These are as follows, in the default descending priority order:
```
```suggestion
The supported goals, in the default descending order of priority, are as follows:
```

PaulRMellor commented on documentation/modules/cruise-control/con-optimization-goals.adoc:
```diff
@@ -30,66 +30,91 @@ For more information on each optimization goal, see link:https://github.com/link
 
 [source]
-RackAwareGoal; ReplicaCapacityGoal; DiskCapacityGoal; NetworkInboundCapacityGoal; NetworkOutboundCapacityGoal; CpuCapacityGoal
+RackAwareGoal; ReplicaCapacityGoal; DiskCapacityGoal; NetworkInboundCapacityGoal; NetworkOutboundCapacityGoal
```
Any reason we don't use a bullet list?

PaulRMellor commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -7,15 +7,17 @@
 
 An _optimization proposal_ is a summary of proposed changes that would produce a more balanced Kafka cluster, with partition workloads distributed more evenly among the brokers. 
-Each optimization proposal is based on the set of xref:con-optimization-goals-{context}[optimization goals] that were used to generate it.
+Each optimization proposal is based on the set of xref:con-optimization-goals-{context}[optimization goals] that was used to generate it, subject to any configured link#capacity-configuration[capacity limits on broker resources].
```
link#capacity-configuration. xref?

PaulRMellor commented on documentation/modules/cruise-control/con-optimization-proposals.adoc:
```diff
@@ -56,7 +58,7 @@ m¦numReplicaMovements
 ¦A measurement of the overall _balancedness_ of a Kafka Cluster, before and after the optimization proposal was generated.
 
-The calculation is 100 minus the sum of the `BalancednessScore` of each violated soft goal. Cruise Control assigns a `BalancednessScore` to every optimization goal based on several factors, including priority (the goal's position in the list of `default.goals` or user-provided goals).
+The score is calculated by subtracting the sum of the `BalancednessScore` of each violated soft goal from 100. Cruise Control assigns a `BalancednessScore` to every optimization goal based on several factors, including priority (the goal's position in the list of `default.goals` or user-provided goals).
```
Maybe the content in brackets doesn't need to be

PaulRMellor commented on documentation/modules/cruise-control/proc-deploying-cruise-control.adoc:
```diff
@@ -89,10 +80,12 @@ spec:
-<2> Defines the Cruise Control configuration, including default and enabled optimization goals. You can provide any xref:ref-cruise-control-configuration-{context}[standard configuration option] apart from those managed directly by {ProductName}.
+<1> Specifies capacity limits for broker resources. For more information, see xref:#capacity-configuration[Capacity configuration].
+<2> Defines the Cruise Control configuration, including the default optimization goals (in `default.goals`) and any customizations to the master optimization goals (in `goals`) or the hard goals (in `hard.goals`). 
+You can provide any xref:ref-cruise-control-configuration-{context}[standard configuration option] apart from those managed directly by {ProductName}. 
```
```suggestion
You can provide any xref:ref-cruise-control-configuration-{context}[standard Cruise Control configuration option] apart from those managed directly by {ProductName}. 
```

PaulRMellor commented on documentation/modules/cruise-control/proc-deploying-cruise-control.adoc:
```diff
@@ -89,10 +80,12 @@ spec:
+For more information on configuring optimization goals, see xref:con-optimization-goals-{context}[]. 
 <3> CPU and memory resources reserved for Cruise Control. For more information, see xref:assembly-resource-limits-and-requests-deployment-configuration-kafka[].
-<4> Defined loggers and log levels added directly (inline) or indirectly (external) through a ConfigMap. A custom ConfigMap must be placed under the log4j.properties key. Cruise Control has a single logger called cruisecontrol.root.logger. You can set the log level to INFO, ERROR, WARN, TRACE, DEBUG, FATAL or OFF. For more information, see xref:logging_configuration[Logging configuration].
+<4> Defined loggers and log levels added directly (inline) or indirectly (external) through a ConfigMap. A custom ConfigMap must be placed under the log4j.properties key. Cruise Control has a single logger called cruisecontrol.root.logger. You can set the log level to INFO, ERROR, WARN, TRACE, DEBUG, FATAL or OFF. For more information, see xref:#logging-configuration[Logging configuration].
```
```suggestion
<4> Defined loggers and log levels added directly (inline) or indirectly (external) through a ConfigMap. A custom ConfigMap must be placed under the log4j.properties key. Cruise Control has a single logger called `cruisecontrol.root.logger`. You can set the log level to INFO, ERROR, WARN, TRACE, DEBUG, FATAL or OFF. For more information, see xref:#logging-configuration[Logging configuration].
```

PaulRMellor commented on documentation/modules/cruise-control/proc-generating-optimization-proposals.adoc:
```diff
@@ -23,7 +23,7 @@ You can then analyze the summary information in the optimization proposal and de
 [source,yaml,subs="attributes+"]
 ----
-apiVersion: {KafkaApiVersion}
+apiVersion: kafka.strimzi.io/v1alpha1
```
Should we have an attribute for this to future-proof?

laidan6000 replied:
I added a new `:KafkaApiVersionAlpha:` attribute.

scholzj replied:
What is the value of having all these as variables? I can understand the value as long as we had all resources using the same API version and upgrading them was as easy as changing the variable. But we have now many CRD and they have their own version lifecycle. So having the variable has no value. You would still need to go through the docs and replace the variables manually.  So I think we should actually drop this from everywhere and have the API version directly in the examples (this is also where #3084 leads)

laidan6000 replied:
Alright, I see your point. I've removed the new variable from this PR.

PaulRMellor commented on documentation/modules/cruise-control/proc-generating-optimization-proposals.adoc:
```diff
@@ -107,7 +109,7 @@ Status:
 ----
 +
-For information about the different properties, see link:#contents-optimization-proposals[Contents of optimization proposals]. 
+For information about the different properties, see xref:#contents-optimization-proposals[Contents of optimization proposals]. 
```
Can we make this clearer: _the different properties_

PaulRMellor commented on documentation/modules/cruise-control/ref-cruise-control-configuration.adoc:
```diff
@@ -92,35 +94,38 @@ spec:
 Since log compaction may remove records needed by Cruise Control, all topics created by Cruise Control must be configured with `cleanup.policy=delete` to disable log compaction.
 Cruise Control will automatically disable log compaction for the `strimzi.cruisecontrol.partitionmetricsamples` and `strimzi.cruisecontrol.modeltrainingsamples` topics.
-The Cruise Control metric reporters will attempt to disable log compaction for the `strimzi.cruisecontrol.metrics` topic but will fail when being started with a new Kafka cluster.
+The Cruise Control metric reporters will attempt to disable log compaction for the `strimzi.cruisecontrol.metrics` topic but will fail if they are started with a new Kafka cluster.
```
```suggestion
The Cruise Control metric reporters will attempt to disable log compaction for the `strimzi.cruisecontrol.metrics` topic, but will fail if they are started with a new Kafka cluster.
```

laidan6000 commented:
Addressed the reviewers' comments. @tomncooper - I left the discussion about master optimization goals open so you can check the note I added.

@tomncooper, @scholzj and @ppatierno - please can you check and approve if happy.

tomncooper commented on documentation/modules/cruise-control/proc-fixing-problems-with-kafkarebalance.adoc:
```diff
@@ -26,13 +26,20 @@ During a “refresh”, a new optimization proposal is requested from the Cruise
++
+[source,shell,subs="+quotes"]
+----
+kubectl get kafka kafkarebalance -o jsonpath='{.status}'
```
I thought we has switched to using `describe`? We should be consistent with whichever command was used elsewhere?

laidan6000 replied:
Well spotted. I changed two more "check status" commands to use `kubectl describe`.
